src/DataValues/DecimalMath.php

```diff
@@ -252,115 +252,62 @@ private function roundDigits( $value, $significantDigits ) {
 			return '+0';
 		}
 
+		$len = strlen( $value );
+
+		// keeping all digits means no rounding
+		if ( $significantDigits >= ( $len -1 ) ) {
+			return $value;
+		}
+
 		if ( $significantDigits < 0 ) {
 			throw new InvalidArgumentException( '$significantDigits must be larger than zero.' );
 		}
 
 		// whether the last character is already part of the integer part of the decimal value
-		$inIntPart = ( strpos( $value, '.' ) === false );
-
-		$rounded = '';
-
-		// Iterate over characters from right to left and build the result back to front.
-		for ( $i = strlen( $value ) -1; $i > 0 && $i > $significantDigits; $i-- ) {
-
-			list( $value, $i, $inIntPart, $next ) = $this->roundNextDigit( $value, $i, $inIntPart );
-
-			$rounded = $next . $rounded;
-		}
-
-		// just keep the remainder of the value as is (this includes the sign)
-		$rounded = substr( $value, 0, $i +1 ) . $rounded;
-
-		if ( strlen( $rounded ) < $significantDigits + 1 ) {
-			if ( $inIntPart ) {
-				$rounded .= '.';
-			}
+		$i = min( $significantDigits +1, $len ); // account for the sign
+		$ch = $i < $len ? $value[$i] : '0';
 
-			$rounded = str_pad( $rounded, $significantDigits+1, '0', STR_PAD_RIGHT );
+		if ( $ch === '.' ) {
+			// NOTE: we expect the input to be well formed, so it cannot end with a '.'
+			$i++;
+			$ch = $i < $len ? $value[$i] : '0';
 		}
 
-		// strip trailing decimal point
-		$rounded = rtrim( $rounded, '.' );
+		// split in significant and insignificant part
+		$rounded = substr( $value, 0, $i );
 
-		return $rounded;
-	}
+		if ( strpos( $rounded, '.' ) === false ) {
+			$suffix = substr( $value, $i );
 
-	/**
-	 * Extracts the next character to add to the result of a rounding run:
-	 * $value[$] will be examined and processed in order to determine the next
-	 * character to prepend to the result (returned in the $nextCharacter field).
-	 *
-	 * Updated values for the parameters are returned as well as the next
-	 * character.
-	 *
-	 * @param string $value
-	 * @param int $i
-	 * @param bool $inIntPart
-	 *
-	 * @return array ( $value, $i, $inIntPart, $nextCharacter )
-	 */
-	private function roundNextDigit( $value, $i, $inIntPart ) {
-		// next digit
-		$ch = $value[$i];
+			// strip insignificant digits after the decimal point
+			$ppos = strpos( $suffix, '.' );
+			if ( $ppos !== false ) {
+				$suffix = substr( $suffix, 0, $ppos );
+			}
 
-		if ( $ch === '.' ) {
-			// just transition from the fractional to the integer part
-			$inIntPart = true;
-			$nextCharacter = '.';
+			// zero out insignificant digits
+			$suffix = strtr( $suffix, '123456789', '000000000' );
```

Comment:
This should be pointless because there is nothing after these zeros that does make a difference. Simply truncate the string.